modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C

```diff
@@ -0,0 +1,166 @@
+
+InputParameters
+NPRBase::validParams()
+{
+  InputParameters params = ElementUserObject::validParams();
+
+  MooseEnum orders("CONSTANT FIRST SECOND THIRD FOURTH");
+  params.addRequiredParam<MooseEnum>(
+      "patch_polynomial_order",
+      orders,
+      "Polynomial order used in least squares fitting of material property "
+      "over the local patch of elements connected to a given node");
+
+  params.addRelationshipManager("ElementSideNeighborLayers",
+                                Moose::RelationshipManagerType::ALGEBRAIC,
+                                [](const InputParameters &, InputParameters & rm_params) {
+                                  rm_params.set<unsigned short>("layers") = 2;
+                                });
+
+  params.addParamNamesToGroup("patch_polynomial_order", "Advanced");
+
+  return params;
+}
+
+NPRBase::NPRBase(const InputParameters & parameters)
+  : ElementUserObject(parameters),
+    _qp(0),
+    _patch_polynomial_order(
+        static_cast<unsigned int>(getParam<MooseEnum>("patch_polynomial_order"))),
+    _multi_index(MathUtils::multiIndex(_mesh.dimension(), _patch_polynomial_order)),
+    _q(_multi_index.size())
+{
+}
+
+Real
+NPRBase::nodalPatchRecovery(const Point & x, const std::vector<dof_id_type> & elem_ids) const
+{
+  // before we go, check if we have enough sample points for solving the least square fitting
+  if (_q_point.size() * elem_ids.size() < _q)
+    mooseError("There are not enough sample points to recover the nodal value, try reducing the "
+               "polynomial order or using a higher-order quadrature scheme.");
+
+  // Assemble the least squares problem over the patch
+  DenseMatrix<Real> A(_q, _q);
+  DenseVector<Real> b(_q);
+  for (auto elem_id : elem_ids)
+  {
+    A += _Ae.at(elem_id);
+    b += _be.at(elem_id);
+  }
+
+  // Solve the least squares fitting
+  DenseVector<Real> coef;
+  A.cholesky_solve(b, coef);
+
+  // Compute the fitted nodal value
+  DenseVector<Real> p = evaluateBasisFunctions(x);
+  return p.dot(coef);
+}
+
+DenseVector<Real>
+NPRBase::evaluateBasisFunctions(Point q_point) const
+{
+  DenseVector<Real> p(_q);
+  Real polynomial;
+  for (unsigned int r = 0; r < _multi_index.size(); r++)
+  {
+    polynomial = 1.0;
+    for (unsigned int c = 0; c < _multi_index[0].size(); c++)
+      for (unsigned int p = 0; p < _multi_index[r][c]; p++)
+        polynomial *= q_point(c);
+    p(r) = polynomial;
+  }
+  return p;
+}
+
+void
+NPRBase::initialize()
+{
+  _Ae.clear();
+  _be.clear();
+}
+
+void
+NPRBase::execute()
+{
+  DenseMatrix<Real> Ae(_q, _q);
+  DenseVector<Real> be(_q);
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+  {
+    DenseVector<Real> p = evaluateBasisFunctions(_q_point[_qp]);
+    DenseMatrix<Real> pxp;
+    pxp.outer_product(p, p);
+    Ae += pxp;
+    be.add(computeValue(), p);
+  }
+
+  dof_id_type elem_id = _current_elem->id();
+  _Ae[elem_id] = Ae;
+  _be[elem_id] = be;
+}
+
+void
+NPRBase::finalize()
```

Comment:
@lindsayad Do you see a type error here? I got the following compile error:
```
/home/gary/projects/moose/modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C:165:62:   required from here
/home/gary/miniconda3/envs/moose/libmesh/include/timpi/standard_type.h:113:19: error: 'TIMPI::StandardType<T, Enable>::StandardType(const T*) [with T = std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > >; Enable = void]' is private within this context
  113 |   StandardType<T> returnval(example);
      |                   ^~~~~~~~~
```
I see that the unspecialized StandardType constructor is private, but why do I hit that?

Reply:
Nevermind, I figured it out. It turns out I need to use a datum that has a nullptr constructor.

Reply:
Can you paste in the whole backtrace? Conceptually I can imagine why there *might* be errors here. The "default" use of `push/pull_parallel_vector_data` is with "fixed" size types, and `DenseMatrix` is a dynamically sized type. However, I've tried to over time to make TIMPI dispatch to appropriate "dynamic" methods from "fixed-size" methods when it's required. At a minimum though, in order for this to work, we must have a `Packing` specialization of `DenseMatrix` and potentially `DenseVector` if you plan to be communicating that around in parallel. @roystgnr have you ever thought about doing `Packing` specializations for these? Off the top of my head I can't think of a barrier to entry here, but can never be too sure whether there are hidden gotchas

Reply:
This is the error from civet [https://civet.inl.gov/job/806916/](https://civet.inl.gov/job/806916/).
I've worked around this using Eigen and passing around `std::vector<Real>`.
```
In file included from /opt/civet/build_2/moose/framework/build/header_symlinks/MooseUtils.h:25,
                 from /opt/civet/build_2/moose/framework/build/header_symlinks/InputParameters.h:13,
                 from /opt/civet/build_2/moose/framework/build/header_symlinks/DistributionInterface.h:12,
                 from /opt/civet/build_2/moose/framework/build/header_symlinks/UserObject.h:13,
                 from /opt/civet/build_2/moose/framework/build/header_symlinks/ElementUserObject.h:13,
                 from /opt/civet/build_2/moose/modules/tensor_mechanics/build/header_symlinks/NPRBase.h:13,
                 from /opt/civet/build_2/moose/modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C:10,
                 from /opt/civet/build_2/moose/modules/tensor_mechanics/build/unity_src/userobjects_nodal_patch_recovery_Unity.C:2:
/opt/civet/build_2/conda_builds/conda_envs/pr-18721/libmesh/include/timpi/standard_type.h: In instantiation of 'TIMPI::StandardType<T> TIMPI::build_standard_type(const T*) [with T = std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > >]':
/opt/civet/build_2/conda_builds/conda_envs/pr-18721/libmesh/include/timpi/parallel_sync.h:383:34:   required from 'void TIMPI::push_parallel_vector_data(const TIMPI::Communicator&, MapToVectors&&, const ActionFunctor&) [with MapToVectors = std::multimap<unsigned int, std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > >, std::less<unsigned int>, std::allocator<std::pair<const unsigned int, std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > > > > >&; ActionFunctor = TIMPI::pull_parallel_vector_data(const TIMPI::Communicator&, const MapToVectors&, GatherFunctor&, ActionFunctor&, const datum*) [with datum = std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > >; MapToVectors = std::unordered_map<unsigned int, std::vector<long unsigned int> >; GatherFunctor = NPRBase::finalize()::<lambda(libMesh::processor_id_type, const std::vector<long unsigned int>&, std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > >&)>; ActionFunctor = NPRBase::finalize()::<lambda(libMesh::processor_id_type, const std::vector<long unsigned int>&, const std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > >&)>]::<lambda(TIMPI::processor_id_type, const std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > >&)>]'
/opt/civet/build_2/conda_builds/conda_envs/pr-18721/libmesh/include/timpi/parallel_sync.h:486:29:   required from 'void TIMPI::pull_parallel_vector_data(const TIMPI::Communicator&, const MapToVectors&, GatherFunctor&, ActionFunctor&, const datum*) [with datum = std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > >; MapToVectors = std::unordered_map<unsigned int, std::vector<long unsigned int> >; GatherFunctor = NPRBase::finalize()::<lambda(libMesh::processor_id_type, const std::vector<long unsigned int>&, std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > >&)>; ActionFunctor = NPRBase::finalize()::<lambda(libMesh::processor_id_type, const std::vector<long unsigned int>&, const std::vector<std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > > >&)>]'
/opt/civet/build_2/moose/modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRBase.C:165:62:   required from here
/opt/civet/build_2/conda_builds/conda_envs/pr-18721/libmesh/include/timpi/standard_type.h:113:19: error: 'TIMPI::StandardType<T, Enable>::StandardType(const T*) [with T = std::pair<long unsigned int, std::pair<libMesh::DenseMatrix<double>, libMesh::DenseVector<double> > >; Enable = void]' is private within this context
  113 |   StandardType<T> returnval(example);
      |                   ^~~~~~~~~
/opt/civet/build_2/conda_builds/conda_envs/pr-18721/libmesh/include/timpi/standard_type.h:85:3: note: declared private here
   85 |   StandardType(const T * example = nullptr);
      |   ^~~~~~~~~~~~
Compiling C++ (in opt mode) /opt/civet/build_2/moose/modules/navier_stokes/build/unity_src/fvbcs_Unity.C...
make: *** [/opt/civet/build_2/moose/framework/build.mk:122: /opt/civet/build_2/moose/modules/tensor_mechanics/build/unity_src/userobjects_nodal_patch_recovery_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
make: *** Waiting for unfinished jobs....
```

Reply:
The only possible catch I can see is the inheritance hierarchy we use with those classes to make the subvectors/submatrices work cleanly.  But I don't think that'll present any obstacle, and there's no other complications I can think of.